bindings/matrix-sdk-ffi/src/client.rs

```diff
@@ -54,18 +60,36 @@ impl Client {
     }
 
     /// Login using a username and password.
-    pub fn login(&self, username: String, password: String) -> anyhow::Result<()> {
+    pub fn login(
+        &self,
+        username: String,
+        password: String,
+        initial_device_name: String,
+        device_id: Option<String>,
+    ) -> anyhow::Result<()> {
         RUNTIME.block_on(async move {
-            self.client.login_username(&username, &password).send().await?;
+            let mut builder = self
+                .client
+                .login_username(&username, &password)
+                .initial_device_display_name(&initial_device_name);
+            let device_id_temp;
+            if device_id.is_some() {
+                device_id_temp = device_id.unwrap();
+                builder = builder.device_id(&device_id_temp)
```

Comment:
This looks a tad weird, can't it be an `if let something` instead?

Reply:
Couldn't do it because the builder method requires a borrowed pointer type.

Reply:
Something like this seems to do the trick on my side:
```
pub fn other_login(
        &self,
        username: String,
        password: String,
        device_id: Option<String>,
    ) -> anyhow::Result<()> {
        RUNTIME.block_on(async move {
            if let Some(device_id) = device_id {
                self.client.login_username(&device_id, &password).send().await?;
            } else {
                self.client.login_username(&username, &password).send().await?;
            }
            Ok(())
        })
    }
```

Reply:
Hmm, yeah it seems to be working for `username` field. I can't tell the difference here but I have tried to do it like below, and i got the compiler error:
```
device_id does not live long enough
borrowed value does not live long enough
```

```
let mut builder = self
    .client
    .login_username(&username, &password)
    .initial_device_display_name(&initial_device_name);
if let Some(device_id) = device_id {
    builder = builder.device_id(&device_id);
}
```

Reply:
Super strange, maybe one of the rust peeps can help us 🤷‍♂️

Reply:
Does an `as_ref()` help here?

```rust
if let Some(device_id) = device_id.as_ref() {
    builder = builder.device_id(device_id);
}
```

`as_ref()` turns you `Option<T>` into a `Option<&T>`.

Reply:
Oh yeah, it helped. Updated 🙏